### ISSUE TITLE
Export fragment annotations of Sirius

### ISSUE THREAD

**Issue by jpfeuffer** [OpenMS/OpenMS#5882]
similar to here but in the basic SiriusAdapter: https://github.com/OpenMS/OpenMS/blob/ddfbb579815265a7273ac99e59605db66555c662/src/utils/AssayGeneratorMetabo.cpp#L467

We can:
- annotate in the original mzML as StringDataArray
- write a new annotated mzML with just the IDed MS2 spectra (easiest)
- annotate in the ID datastructure together with the reference to the spectrum (best)

Note: An identification could have multiple originating spectra, if feature info was given and MS2 spectra inside a feature were merged.

**Comment by oliveralka** [comment]
I think a lot of preparation in terms of retrieving the annotation and storing it a a spectrum is already done here:
https://github.com/OpenMS/OpenMS/blob/develop/src/tests/class_tests/openms/source/SiriusFragmentAnnotation_test.cpp
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/FORMAT/DATAACCESS/SiriusFragmentAnnotation.cpp#L47

The `annotated_msspectrum.getNativeID()` stores the information about all spectra used for a specific feature.

**Comment by jpfeuffer** [comment]
Yep, the question is, if we want to store it together with the spectrum, since it actually belongs more to the ID/ the compound-spectrum match. We would need to check how to encode it in one of the two ID structures. The idea would be to be able to click through different candidates and compare which/how many fragments match the spectrum.
The problem with the the old ID structure is of course that it does not have a notion of multiple Origin spectra in case of merging. We might need to export merged spectra as well to reference them.

**Comment by jpfeuffer** [comment]
@oliveralka are the results of Sirius enough to reconstruct the whole spectrum?
E.g. https://github.com/OpenMS/OpenMS/blob/master/src/openms/include/OpenMS/FORMAT/DATAACCESS/SiriusFragmentAnnotation.h#L71-L71

returns completely new spectra from the annotations that result from Sirius. But what I want to do is to give in the spectra **as they were** and only **add** annotations.
This only works if Sirius provides all peaks, even if they are not annotated (which sounds unlikely).

**Comment by jpfeuffer** [comment]
Oh and can you quickly point me to where the merged spectra are passed to sirius?
I think I will just write them out with annotations. Too much trouble to encode that with the identifications.

**Comment by oliveralka** [comment]
Unfortunately, there are no merged spectra passed to SIRIUS. All spectra for a specific feature are passed via the SiriusMSFile, which is the SIRIUS internal format. SIRIUS merges them internally. 

Example .ms file:
https://github.com/OpenMS/OpenMS/blob/develop/src/tests/class_tests/openms/data/SiriusFragmentAnnotation_test/spectrum.ms

One way to get some of the data would be the "feature_mapping" output of https://github.com/OpenMS/OpenMS/blob/develop/src/utils/SiriusAdapter.cpp#L186

I have a meeting with Markus regarding the new SIRIUS later this week. I will ask him if there is a way to get the non-annotated information. You could potentially reconstruct the original spectrum from the ">ms2peaks" given specturm.ms file, but I am not sure if that is what you want.

**Comment by jpfeuffer** [comment]
ok hmm well, I guess until we can access the merged spectrum, we could annotate all spectra with the same result annotations right? Then you could choose which spectrum to look at.
How do I get the info which original spectra went into a specific sirius output? I saw somewhere the nativeIDs are concatenated for a file/subfolder? Is that right?

**Comment by oliveralka** [comment]
The concatenated nativeIDs are set as Native ID in the SiriusFragmentAnnotation.
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/FORMAT/DATAACCESS/SiriusFragmentAnnotation.cpp#L108

Yes, they correspond to the original nativeID in the mzML file. 
Yes, ConsensusXML is not yet supported.

**Comment by jpfeuffer** [comment]
Hi @oliveralka 
So far so good. I can export the annotated peaks to an mzML now.

One question: What is this line for?
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/FORMAT/DATAACCESS/SiriusFragmentAnnotation.cpp#L73

Why do you need to make the spectra unique in that regard?
I thought about having one "annotation spectrum" _per_ candidate _per_ original spectrum. Am I missing something?

**Comment by jpfeuffer** [comment]
Hmm I am not sure I understand. I thought your code only extracts the first ranking candidate anyway. How can you have multiple first ranks for the same feature?

**Comment by oliveralka** [comment]
Yes, that is true, but it also resolves ambiguities introduced by multiple identifications for the same feature. 
Let's say you have one feature which has two possible AMS identifications (`ID1[C19H26O6,[M+Na]+]`; `ID2[C14H17IN2O2,[M+H]+]`) then SIRIUS runs for each identification. Only one of these can be correct, so I use the SIRIUS score to assess which one - this is the one which is reported.

**Comment by jpfeuffer** [comment]
Ah I see. But in that case there would never be multiple candidates reported by Sirius right? In that mode it only evaluates the one formula given to it right?

Wouldn't it be cool then to be able to give it ALL the possible candidates and it runs as normal ("unknown") and ranks the given candidates?

Then we only would need one level of resolution. But I guess this is not possible as a mode in Sirius yet?

**Comment by jpfeuffer** [comment]
By the way, do you know if you get SMILES annotations of peaks after CSI:FingerID? Is there an example output after fingerid somewhere?

**Comment by eeko-kon** [comment]
Yes you do, when it finds matches to the spectral libraries (pubchem, etc). Example output from SiriusAdapter using CSI:
[structures_Epemicins.csv](https://github.com/OpenMS/OpenMS/files/8304643/structures_Epemicins.csv)

**Comment by jpfeuffer** [comment]
Ah! Thanks! But this is for the whole spectrum, right? I am currently investigating the possibility to annotate the peaks in the fragment spectrum with the output from Sirius.

**Comment by oliveralka** [comment]
> Ah I see. But in that case there would never be multiple candidates reported by Sirius right? In that mode it only evaluates the one formula given to it right?

As far as I know - yes that is correct.

> Wouldn't it be cool then to be able to give it ALL the possible candidates and it runs as normal ("unknown") and ranks the given candidates? Then we only would need one level of resolution. But I guess this is not possible as a mode in Sirius yet?

Yes, that would actually be the best, but this is currently not possible. I will ask Markus. From our side the current implementation (how the identifications are passed to SIRIUS) would have to be changed as well, since each one is written as a .ms entry. 

>By the way, do you know if you get SMILES annotations of peaks after CSI:FingerID? Is there an example output after fingerid somewhere?

No, the fragment annotation is performed on SIRIUS level. There are no SMILES annotations available at peak level even after CSIFingerID was run. 

Here is a zip archive of the SIRIUS poject space based on the test SiriusAdapter_9 (with some minor changes)

```
${TOPP_BIN_PATH}/SiriusAdapter -test -sirius_executable "${SIRIUS_BINARY}" -in ${DATA_DIR_TOPP}/AssayGeneratorMetabo_decoy_generation_input.mzML  -in_featureinfo ${DATA_DIR_TOPP}/AssayGeneratorMetabo_decoy_generation_input_multids.featureXML -out_sirius SiriusAdapter_9_output.tmp -out_fingerid bla.tmp -sirius:profile qtof -sirius:db all -project:processors 1 -out_project_space "/home/pancake/software/test_project_space")
```

[Sirius_out_SA_9_withProjectSpace_andFingerID.zip](https://github.com/OpenMS/OpenMS/files/8306548/Sirius_out_SA_9_withProjectSpace_andFingerID.zip)

**Comment by jpfeuffer** [comment]
I implemented it mostly in a recent PR.
